indexing-service/src/main/java/org/apache/druid/indexing/common/actions/SegmentNukeAction.java

```diff
@@ -54,26 +56,33 @@ public Set<DataSegment> getSegments()
   }
 
   @Override
-  public TypeReference<Void> getReturnTypeReference()
+  public TypeReference<Set<DataSegment>> getReturnTypeReference()
   {
-    return new TypeReference<Void>()
+    return new TypeReference<Set<DataSegment>>()
     {
     };
   }
 
   @Override
-  public Void perform(Task task, TaskActionToolbox toolbox)
+  public Set<DataSegment> perform(Task task, TaskActionToolbox toolbox)
   {
     TaskLocks.checkLockCoversSegments(task, toolbox.getTaskLockbox(), segments);
-
+    final Set<DataSegment> segmentsToKill = new HashSet<>();
     try {
       toolbox.getTaskLockbox().doInCriticalSection(
           task,
           segments.stream().map(DataSegment::getInterval).collect(Collectors.toSet()),
           CriticalAction.builder()
                         .onValidLocks(
                             () -> {
-                              toolbox.getIndexerMetadataStorageCoordinator().deleteSegments(segments);
+                              final IndexerMetadataStorageCoordinator coordinator
+                                  = toolbox.getIndexerMetadataStorageCoordinator();
+
+                              // Find subset of segments with unreferenced load specs assuming successful nuking
+                              segmentsToKill.addAll(coordinator.findSegmentsWithUnreferencedLoadSpecs(segments));
+
```

Comment:
Design wise, it seems strange that the nuke task action is doing this. Shouldn't the kill task be responsible for determining this set or perhaps a separate task action?

If there's a solid rationale for doing this, can you please update the javadocs so someone reading the code isn't caught by surprise.

Reply:
Separated it to a new task action